typed-racket-lib/typed-racket/typecheck/tc-structs.rkt

```diff
@@ -234,158 +238,118 @@
     (for*/and ([var (in-list tvars)]
                [t (in-list fields)])
       (let ([variance (hash-ref (free-vars-hash (free-vars* t)) var variance:const)])
         (or (variance:const? variance)
             (and (not mutable) (variance:co? variance))))))
+
   (define covariant?
     (and (covariant-for? self-fields mutable)
          (covariant-for? parent-fields parent-mutable)))
 
   (define (poly-wrapper t) (make-Poly tvars t))
-  (define bindings
-    (list*
-     ;; the list of names w/ types
-     (make-def-binding (struct-names-struct-type names) (make-StructType sty))
-     (make-def-binding (struct-names-predicate names)
-                       (make-pred-ty (if (not covariant?)
-                                         (make-StructTop sty)
-                                         (subst-all (make-simple-substitution
-                                                     tvars (map (const Univ) tvars)) poly-base))))
-     (append
-      (for/list ([g (in-list (struct-names-getters names))]
-                 [t (in-list self-fields)]
-                 [i (in-naturals parent-count)])
-        (let* ([path (make-StructPE poly-base i)]
-               [func (poly-wrapper
-                      (if mutable
-                          (->* (list poly-base) t)
-                          (->acc (list poly-base) t (list path))))])
-          (add-struct-accessor-fn! g poly-base i mutable #f)
-          (make-def-binding g func)))
-      (if mutable
-          (for/list ([s (in-list (struct-names-setters names))]
-                     [t (in-list self-fields)]
-                     [i (in-naturals parent-count)])
-            (add-struct-mutator-fn! s poly-base i #f)
-            (make-def-binding s (poly-wrapper (->* (list poly-base t) -Void))))
-          null))))
+  (mk-bindings! sty si names desc
+                (lambda (poly-base)
+                  (if (not covariant?)
+                      (make-StructTop sty)
+                      (subst-all (make-simple-substitution
+                                  tvars (map (const Univ) tvars)) poly-base)))
+                (lambda (name self-fld idx-parent-cnt poly-base)
+                  (let* ([path (make-StructPE poly-base idx-parent-cnt)]
+                         [func (poly-wrapper
+                                (if mutable
+                                    (->* (list poly-base) self-fld)
+                                    (->acc (list poly-base) self-fld (list path))))])
+                    (values (list poly-base idx-parent-cnt mutable #f)
+                            func)))
+                (lambda (_ self-fld idx-parent-cnt poly-base)
+                  (values (list poly-base idx-parent-cnt #f)
+                          (poly-wrapper (->* (list poly-base self-fld) -Void))))))
+
+
+(define-syntax-rule (nullable condition expr)
+  (if condition
+      expr
+      null))
+
+(define (mk-bindings! sty si names desc mk-pred-ty mk-getter-vals mk-setter-vals)
```

Comment:
signature/purpose/contract

this seems like the important one, so explaining this well is what makes the refactoring worth it.